services/QuillLMS/app/models/teacher_notification.rb

```diff
@@ -0,0 +1,68 @@
+# frozen_string_literal: true
+
+# == Schema Information
+#
+# Table name: teacher_notifications
+#
+#  id                :bigint           not null, primary key
+#  email_sent        :datetime
+#  notification_type :text             not null
+#  params            :jsonb
+#  created_at        :datetime         not null
+#  updated_at        :datetime         not null
+#  user_id           :bigint           not null
+#
+# Indexes
+#
+#  index_teacher_notifications_on_user_id  (user_id)
+#
+# Foreign Keys
+#
+#  fk_rails_...  (user_id => users.id)
+#
+class TeacherNotification < ApplicationRecord
+  NOTIFICATION_TYPES = [
+    STUDENT_COMPLETED_DIAGNOSTIC = 'student-completed-diganostic',
+    STUDENT_COMPLETED_ALL_DIAGNOSTIC_RECOMMENDATIONS = 'student-completed-all-diagnostic-recommendations',
+    STUDENT_COMPLETED_ALL_ASSIGNED_ACTIVITIES = 'student-completed-all-assigned-activities'
+  ]
+
+  NOTIFICATION_TYPE_VALIDATION = {
+    STUDENT_COMPLETED_DIAGNOSTIC => [
+      :student_name,
+      :classroom_name,
+      :diagnostic_name
+    ],
+    STUDENT_COMPLETED_ALL_DIAGNOSTIC_RECOMMENDATIONS => [
+      :student_name,
+      :classroom_name
+    ],
+    STUDENT_COMPLETED_ALL_ASSIGNED_ACTIVITIES => [
+      :student_name,
+      :classroom_name
+    ]
+  }
+
+  belongs_to :user
+
+  validates :notification_type, presence: true, inclusion: {in: NOTIFICATION_TYPES}
+  validate :params_must_be_hash, :params_payload_has_required_keys
+
+  private def params_is_hash?
+    params.is_a?(Hash)
```

Comment:
Since `params` has special meaning in rails in the context of the controller, to avoid confusion later, I would name this something more descriptive (e.g. custom_attrs).